lib/_stream_readable.js

```diff
@@ -570,7 +570,8 @@ function maybeReadMore(stream, state) {
 function maybeReadMore_(stream, state) {
   var len = state.length;
   while (!state.reading && !state.ended &&
-         state.length < state.highWaterMark) {
+         (state.length < state.highWaterMark ||
+          (state.flowing && state.length === 0))) {
```

Comment:
I'm not convinced by this change. In the past there was a check on `!state.flowing`, see https://github.com/nodejs/node/commit/1e0f3315c77033ef0e01bb37c3d41c8e1d65e686#diff-ba6a0df0f5212f5cba5ca5179e209a17L540.

Can you please add a comment on top of this while loop explaining the conditions?

Reply:
That commit might be the specific one that broke the `highWaterMark: 0` case. Previously the `read(0)` was called immediately after `emit('data', ...)`. In that commit the `read(0)` call was removed and the while-loop now executes in the flowing mode.

A side effect of this is that the `state.length` is now not _only_ buffer, but also the 'first bit of data emitted immediately', which is why we shouldn't terminate the loop in `highWaterMark === 0` scenario when the stream is flowing.

I'll add the comment when I get to my PC later today.

Reply:
Added the comment (and amended it a bit, sorry for the --force-with-lease).

I started wondering whether the `state.length === 0` is a superfluous check in case the stream is flowing - but it seems `_read()` is called in this case even if the stream is already above `highWaterMark` so it's probably safer to just keep that check in there.